File: components/breadcrumb/android/CiataBreadcrumb.kt
package org.ciata.ds.breadcrumb

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp

/** Implementação experimental do CMP-0016 Breadcrumb para Jetpack Compose. */
@Composable
fun CiataBreadcrumb(
    items: List<Pair<String, (() -> Unit)?>>,
) {
    require(items.isNotEmpty()) { "items não pode ser vazio." }

    val normalizedItems = items.map { (label, action) -> label.trim() to action }
    require(normalizedItems.all { it.first.isNotEmpty() }) { "rótulos não podem ser vazios." }
    require(normalizedItems.last().second == null) { "o item atual deve ser não interativo." }
    require(normalizedItems.dropLast(1).all { it.second != null }) {
        "itens anteriores ao atual devem ser interativos."
    }

    FlowRow(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        normalizedItems.forEach { (label, action) ->
            if (action == null) {
                Text(label)
            } else {
                TextButton(onClick = action) {
                    Text(label)
                }
            }
        }
    }
}

File: components/breadcrumb/ios/CiataBreadcrumb.swift
import SwiftUI

/// Implementação experimental do CMP-0016 Breadcrumb para SwiftUI.
public struct CiataBreadcrumb: View {
    public struct Item: Identifiable {
        public let id = UUID()
        public let label: String
        public let action: (() -> Void)?

        public init(label: String, action: (() -> Void)? = nil) {
            let normalizedLabel = label.trimmingCharacters(in: .whitespacesAndNewlines)
            precondition(!normalizedLabel.isEmpty, "label não pode ser vazio")
            self.label = normalizedLabel
            self.action = action
        }
    }

    private let items: [Item]

    public init(items: [Item]) {
        precondition(!items.isEmpty, "items não pode ser vazio")
        precondition(items.last?.action == nil, "o item atual deve ser não interativo")
        precondition(items.dropLast().allSatisfy { $0.action != nil }, "itens anteriores ao atual devem ser interativos")
        self.items = items
    }

    public var body: some View {
        HStack(spacing: 8) {
            ForEach(items) { item in
                if let action = item.action {
                    Button(item.label, action: action)
                } else {
                    Text(item.label)
                }
            }
        }
    }
}
